File: apps/android/app/src/main/java/ai/openclaw/app/gateway/GatewayCustomHeaders.kt
package ai.openclaw.app.gateway

import ai.openclaw.mobile.core.GatewayCustomHeaders as SharedGatewayCustomHeaders
import ai.openclaw.mobile.core.HeaderValuePolicy

/**
 * Operator-defined HTTP headers attached to gateway connections so gateways fronted by
 * authenticating reverse proxies (Cloudflare Access-style service tokens) stay reachable.
 * Header values are credentials: persist them only in SecurePrefs and never log them.
 */
object GatewayCustomHeaders {
  fun isReservedName(name: String): Boolean = SharedGatewayCustomHeaders.isReservedName(name)

  /**
   * Drops entries that cannot travel as a single well-formed header: empty, reserved, or
   * non-token names, and values outside printable ASCII. Dropping invalid entries keeps one bad
   * stored value from wedging every reconnect or being interpreted differently by a proxy.
   */
  fun sanitized(headers: Map<String, String>): Map<String, String> =
    SharedGatewayCustomHeaders.sanitized(headers, HeaderValuePolicy.ASCII_PRINTABLE)
}

File: apps/shared/OpenClawKit/Sources/OpenClawKit/GatewayCustomHeaders.swift
import Foundation
import OpenClawMobileCore

/// Operator-defined HTTP headers attached to every gateway-bound request so gateways fronted
/// by authenticating reverse proxies (Cloudflare Access-style service tokens) stay reachable.
/// Header values are credentials: persist them in the platform secure store and never log them.
public enum GatewayCustomHeaders {
    public static func isReservedName(_ name: String) -> Bool {
        MobileCoreBridge.shared.isReservedGatewayHeaderName(name: name)
    }

    /// Drops entries that cannot travel as a single well-formed header: empty, reserved, or
    /// non-token names, and values with control characters (request-splitting guard).
    public static func sanitized(_ headers: [String: String]) -> [String: String] {
        guard JSONSerialization.isValidJSONObject(headers),
              let input = try? JSONSerialization.data(withJSONObject: headers),
              let inputJSON = String(data: input, encoding: .utf8),
              let output = MobileCoreBridge.shared.sanitizeGatewayHeadersForApple(headersJson: inputJSON).data(using: .utf8),
              let sanitized = try? JSONSerialization.jsonObject(with: output) as? [String: String]
        else {
            return [:]
        }
        return sanitized
    }
}
